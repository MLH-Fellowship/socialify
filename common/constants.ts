// GitHub constants.
export const GITHUB_GRAPHQL_ENDPOINT: string = 'https://api.github.com/graphql'

// Socialify constants.
export const SOCIALIFY_GRAPHQL_ENDPOINT: string = '/api/graphql'
