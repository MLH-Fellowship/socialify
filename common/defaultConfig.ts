import { repoQueryResponse } from './relay/__generated__/repoQuery.graphql'
import Configuration, {
  FileType,
  Font,
  OptionalConfigs,
  Pattern,
  Theme
} from './types/configType'

const getOptionalConfig = (repository: repoQueryResponse['repository']) => {
  if (repository) {
    const languages = repository.languages?.nodes || []
    const language =
      languages.length > 0 ? languages[0]?.name || 'unknown' : 'unknown'
    const newConfig: OptionalConfigs = {
      owner: { state: true, value: repository.owner.login },
      description: {
        state: false,
        editable: true,
        value: repository.description || ''
      },
      language: { state: true, value: language },
      stargazers: { state: true, value: repository.stargazerCount },
      forks: { state: false, value: repository.forkCount },
      pulls: { state: false, value: repository.pullRequests.totalCount },
      issues: { state: false, value: repository.issues.totalCount }
    }
    return newConfig
  }
  return null
}

const defaultConfig: Configuration = {
  name: '',
  logo: '',
  font: Font.inter,
  theme: Theme.dark,
  pattern: Pattern.plus,
  fileType: FileType.png
}

export { defaultConfig, getOptionalConfig }
