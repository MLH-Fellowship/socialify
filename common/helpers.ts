import { Pattern, Theme } from './types/configType'
import {
  signal,
  charlieBrown,
  formalInvitation,
  plus,
  circuitBoard,
  overlappingHexagons,
  brickWall,
  floatingCogs,
  diagonalStripes
} from 'hero-patterns'

const getDevIconClassName = (language: string, theme: Theme): string => {
  const LANGUAGE_ICON_MAPPING: { [key: string]: string } = {
    C: 'c-plain',
    'C#': 'csharp-plain',
    'C++': 'cplusplus-plain',
    CoffeeScript: 'coffeescript-original',
    CSS: 'css3-plain',
    Go: 'go-plain',
    Groovy: 'groovy-plain',
    HTML: 'html5-plain',
    Java: 'java-plain',
    JavaScript: 'javascript-plain',
    'Jupyter Notebook': 'python-plain',
    PHP: 'php-plain',
    Python: 'python-plain',
    Ruby: 'ruby-plain',
    Rust: 'rust-plain',
    Scala: 'scala-plain',
    Swift: 'swift-plain',
    TypeScript: 'typescript-plain',
    GitHub: 'github-original',
    DevIcon: 'devicon-plain'
  }

  return `devicon-${LANGUAGE_ICON_MAPPING[language] || 'devicon-plain'} ${
    theme === Theme.light ? 'colored' : ''
  }`
}

const getHeroPattern = (pattern: Pattern, theme: Theme): string => {
  const PATTERN_FUNCTIONS_MAPPING: { [key: string]: any } = {
    [Pattern.signal]: signal,
    [Pattern.charlieBrown]: charlieBrown,
    [Pattern.formalInvitation]: formalInvitation,
    [Pattern.plus]: plus,
    [Pattern.circuitBoard]: circuitBoard,
    [Pattern.overlappingHexagons]: overlappingHexagons,
    [Pattern.brickWall]: brickWall,
    [Pattern.floatingCogs]: floatingCogs,
    [Pattern.diagonalStripes]: diagonalStripes,
    [Pattern.solid]: null
  }
  const patternFunction = PATTERN_FUNCTIONS_MAPPING[pattern]
  const themedBackgroundColor = theme === Theme.dark ? '#000' : '#fff'

  if (!patternFunction) return themedBackgroundColor

  const darkThemeArgs = ['#eaeaea', 0.2]
  const lightThemeArgs = ['#eaeaea', 0.6]
  const patternImageUrl = patternFunction.apply(
    null,
    theme === Theme.dark ? darkThemeArgs : lightThemeArgs
  )

  return `${themedBackgroundColor} ${patternImageUrl}`
}

let webpSupport: boolean | undefined
const checkWebpSupport = (): boolean => {
  if (webpSupport !== undefined) {
    return webpSupport
  }

  webpSupport = (() => {
    try {
      const canvas = document.createElement('canvas')
      return canvas.toDataURL('image/webp').indexOf('data:image/webp') === 0
    } catch (e) {
      return false
    }
  })()

  return webpSupport
}

const HOST_PREFIX = process.env.VERCEL_URL
  ? `https://${process.env.VERCEL_URL}`
  : process.env.PROJECT_URL || ''

export { getDevIconClassName, getHeroPattern, checkWebpSupport, HOST_PREFIX }
