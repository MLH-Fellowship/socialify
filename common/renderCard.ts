import { readFileSync } from 'fs'
import React from 'react'
import ReactDOMServer from 'react-dom/server'
import { flushToHTML } from 'styled-jsx/server'
import { fetchQuery } from 'react-relay'

import Card from '../src/components/preview/card'
import enviornment from './relay/environment'
import repoQuery from './relay/repoQuery'

import { repoQueryResponse } from './relay/__generated__/repoQuery.graphql'

import { Font } from './types/configType'
import QueryType from './types/queryType'

import { mergeConfig } from './configHelper'

const cwd = process.cwd()

const devIconCSS = readFileSync(`${cwd}/common/fonts/devicon.css`).toString(
  'utf-8'
)

const getGoogleFontCSS = (font: Font): string => {
  const googleFontsCSS = readFileSync(`${cwd}/.next/google-fonts.css`).toString(
    'utf-8'
  )

  return googleFontsCSS
    .replace(/([{;])\n*\s*/g, '$1')
    .split('\n')
    .filter(f => f.startsWith(`@font-face {font-family: '${font}'`))
    .join('\n')
}

const getRepoResponse = async (owner: string, name: string) => {
  return (await fetchQuery(enviornment, repoQuery, {
    owner,
    name
  })) as repoQueryResponse
}

export default async (query: QueryType) => {
  const { repository } = await getRepoResponse(query._owner, query._name)
  const config = mergeConfig(repository, query)

  if (!config) throw Error('Configuration failed to generate')

  const cardComponent = React.createElement(Card, config)
  const cardHTMLMarkup = ReactDOMServer.renderToStaticMarkup(cardComponent)
  const styleTags = flushToHTML()

  return `<!DOCTYPE html>
<html>
  <meta charset="utf-8">
  <title>Generated Image</title>
  <link rel="icon" href="data:,">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <style>
    ${devIconCSS}
    ${getGoogleFontCSS(config.font)}
  </style>
  ${styleTags}
  <body>
    ${cardHTMLMarkup}
  </body>
</html>`
}
