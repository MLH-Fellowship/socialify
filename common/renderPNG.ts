import React from 'react'
import { ImageResponse } from '@vercel/og'
import Card from '../src/components/preview/card'
import { getCardConfig, getFonts } from './renderCard'
import QueryType from './types/queryType'

const renderCardPNG = async (query: QueryType) => {
  const config = await getCardConfig(query)

  return new ImageResponse(React.createElement(Card, config), {
    width: 1280,
    height: 640,
    fonts: await getFonts(config.font),
    emoji: 'twemoji'
  })
}

export default renderCardPNG
