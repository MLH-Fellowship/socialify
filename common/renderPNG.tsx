import { ImageResponse } from '@vercel/og'
import Card from '../src/components/preview/card'
import { getCardConfig, getFonts } from './renderCard'
import QueryType from './types/queryType'

const renderCardPNG = async (query: QueryType) => {
  const config = await getCardConfig(query)

  return new ImageResponse(<Card {...config} />, {
    width: 1280,
    height: 640,
    fonts: await getFonts(config.font),
    emoji: 'twemoji'
  })
}

export default renderCardPNG
