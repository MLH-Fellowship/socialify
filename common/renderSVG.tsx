// @ts-ignore
import satori, { init as initSatori } from 'satori/wasm'
// @ts-ignore
import initYoga from 'yoga-wasm-web'
// @ts-ignore
import yogaWasm from '../public/yoga.wasm?module'

import Card from '../src/components/preview/card'
import { getCardConfig, getFonts, loadDynamicAsset } from './renderCard'
import QueryType from './types/queryType'

const renderCardSVG = async (query: QueryType) => {
  const yoga = await initYoga(yogaWasm)
  initSatori(yoga)

  const config = await getCardConfig(query)

  return satori(<Card {...config} />, {
    width: 1280,
    height: 640,
    fonts: await getFonts(config.font),
    loadAdditionalAsset: loadDynamicAsset
  })
}

export default renderCardSVG
