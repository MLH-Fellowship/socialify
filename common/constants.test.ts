import * as constants from '@/common/constants'

describe('constants', () => {
  test('matches snapshot', () => {
    expect(constants).toMatchSnapshot()
  })
})
