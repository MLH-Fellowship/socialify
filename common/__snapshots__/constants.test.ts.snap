// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constants matches snapshot 1`] = `
{
  "GITHUB_GRAPHQL_ENDPOINT": "https://api.github.com/graphql",
  "SOCIALIFY_GRAPHQL_ENDPOINT": "/api/graphql",
}
`;
